pkg/menus/menus.go

```diff
@@ -0,0 +1,174 @@
+	}
+
+	if choose == "" {
+		for _, en := range entries {
+			if en.IsDefault() {
+				return en, nil
+			}
+		}
+	} else {
+		c, err := strconv.Atoi(choose)
+		if err != nil {
+			return nil, fmt.Errorf("Error at convert input to number in desplay menu: %v", err)
+		}
+		return entries[c-1], nil
+	}
+	return nil, nil
+}
+
+// GetInput will present an input box to user and return the user's input.
+// GetInout will check validation of input using checkValidFunc.
+func GetInput(introwords string, location int, wid int, ht int, checkValidFunc func(string) (bool, string)) (string, error) {
+	// intro paragraph is to tell user what need to be input here
+	intro := widgets.NewParagraph()
+	intro.Text = introwords
+	intro.Border = false
+	intro.SetRect(0, location, len(introwords)+4, location+3)
+	intro.TextStyle.Fg = ui.ColorWhite
+	ui.Render(intro)
+	location += 3
+
+	// input paragraph is where user input
+	input := widgets.NewParagraph()
+	input.Text = ""
+	intro.Border = false
+	input.SetRect(0, location, wid, location+ht+2)
+	input.TextStyle.Fg = ui.ColorWhite
+	ui.Render(input)
+	location += ht + 2
+
+	// warning paragraph is to warn user their input is not valid
+	warning := widgets.NewParagraph()
+	warning.Text = ""
+	warning.Border = false
+	warning.SetRect(0, location, wid, location+3)
+	warning.TextStyle.Fg = ui.ColorWhite
+	ui.Render(warning)
+
+	// keep tracking all input from user
+	uiEvents := ui.PollEvents()
+	for {
+		e := <-uiEvents
+		if e.Type != ui.KeyboardEvent {
+			continue
+		}
+		switch e.ID {
+		// if user input q or control-c, exit the program.
+		case "q", "<C-c>":
+			ui.Close()
+			ui.Close()
+			os.Exit(1)
+		// If user hit enter means he did his choice.
+		// So check whether the input is valid or not by checkVilidFunc function.
+		// If the input is not acceptable, show a warning.
+		case "<Enter>":
+			valid, warningWords := checkValidFunc(input.Text)
+			// if input is vilid, directly return the input itself.
+			// else we should clear the input box and show warnings.
+			if valid {
+				return input.Text, nil
+			}
+			input.Text = ""
+			warning.Text = warningWords
+			ui.Render(input)
+			ui.Render(warning)
+		//If userhit backspace, remove the last input character
+		case "<Backspace>":
+			input.Text = input.Text[:len(input.Text)-1]
+			ui.Render(input)
+		// as long as user do not hit enter or q or control-c, assuming that user is still inputing his choice
+		default:
+			// clear warning when user input
+			if warning.Text != "" {
+				warning.Text = ""
+				ui.Render(warning)
+			}
+			input.Text += e.ID
+			ui.Render(input)
+		}
+	}
+}
+
+// NewInputWindow will create a new ui window and display an input box.
+// We have this function because GetInput function will display in the current window.
+// And sometimes we want anew window to make the UI looks clean and neat.
+func NewInputWindow(introwords string, wid int, ht int, checkValidFunc func(string) (bool, string)) (string, error) {
```

Comment:
It seems that you always call NewInputWindow with the same wid and ht. What do you think of removing those arguments and doing something like

```
func NewInputWindow(desc string, isValid func(string) (bool, string)) (string, error) {
  return NewCustomInputWindow(desc, 100, 1, isValid)
}

func NewCustomInputWindow(desc string, wid, ht int, isValid func(string) (bool, string)) (string, error) {
}
```

so that way you don't need to  keep inputting widt/ht if you can use the default, and you have the option to have a custom size if you want